File: scripts/generateSyntheticData.ts
import { faker } from '@faker-js/faker';
import { writeFile } from 'fs/promises';

interface BugReport {
  fullName: string;
  email: string;
  feedbackType: 'Bug' | 'Feature' | 'UI Issue';
  appVersion: string;
  stepsToReproduce: string;
  expectedBehavior: string;
  actualBehavior: string;
  screenshotProvided: boolean;
  label: 'complete' | 'incomplete' | 'vague' | 'invalid';
}

const FEEDBACK_TYPES = ['Bug', 'Feature', 'UI Issue'] as const;
const DEVICES = ['Android 13', 'iOS 17', 'Windows 11', 'macOS 14', 'Ubuntu 22.04'];
const ACTIONS = [
  'tap the Save button',
  'open the settings page',
  'log out',
  'upload a file',
  'enter text in the search bar',
  'start the app',
  'switch accounts',
  'change the theme',
  'select a menu item',
  'resize the window'
];
const ISSUES = [
  'crashes with error code 500',
  'freezes after login',
  'shows a blank screen',
  'does not respond',
  'crashed imediately',
  'bttn not responsive',
  'screen go blank',
  'resets the form',
  'closes unexpectedly',
  'shows garbled text'
];

function shuffle<T>(arr: T[]): T[] {
  return faker.helpers.shuffle(arr);
}

function validVersion(): string {
  const major = faker.number.int({ min: 0, max: 5 });
  const minor = faker.number.int({ min: 0, max: 9 });
  const patch = faker.number.int({ min: 0, max: 9 });
  const suffix = faker.helpers.arrayElement(['', '-beta', '-rc1']);
  return `v${major}.${minor}.${patch}${suffix}`;
}

function invalidVersion(): string {
  return faker.helpers.arrayElement([
    'v1',
    '1.0',
    'version' + faker.number.int({ min: 6, max: 99 }),
    `v${faker.word.words(1)}`,
    ''
  ]);
}

function sentenceWithIssue(): string {
  const device = faker.helpers.arrayElement(DEVICES);
  const action = faker.helpers.arrayElement(ACTIONS);
  const issue = faker.helpers.arrayElement(ISSUES);
  return `On ${device}, when I ${action}, the app ${issue}.`;
}

function shortIssue(): string {
  return faker.helpers.arrayElement([
    'app crashed',
    'it froze',
    'something broke',
    'not sure',
    ''
  ]);
}

function generateEntry(label: BugReport['label']): BugReport {
  const fullName = label === 'invalid' && Math.random() < 0.3
    ? faker.word.words(2)
    : faker.person.fullName();

  const email = label === 'invalid'
    ? faker.word.words(1) + '.com'
    : label === 'incomplete' && Math.random() < 0.3
      ? ''
      : faker.internet.email();

  const feedbackType = faker.helpers.arrayElement(FEEDBACK_TYPES);
  const screenshotProvided = faker.datatype.boolean();

  let appVersion = validVersion();
  let steps = sentenceWithIssue();
  let expected = `The app should ${faker.word.verb()} without errors.`;
  let actual = `Instead, it ${faker.helpers.arrayElement(ISSUES)}.`;

  if (label === 'vague') {
    steps = shortIssue();
    expected = faker.helpers.arrayElement(['should work', 'should not crash', '']);
    actual = faker.helpers.arrayElement(['didn\'t', 'crashed', '']);
  } else if (label === 'incomplete') {
    appVersion = Math.random() < 0.5 ? validVersion() : invalidVersion();
    steps = Math.random() < 0.5 ? shortIssue() : '';
    expected = Math.random() < 0.5 ? '' : 'should work';
    actual = Math.random() < 0.5 ? '' : faker.helpers.arrayElement(['didn\'t', 'crashed']);
  } else if (label === 'invalid') {
    appVersion = invalidVersion();
    steps = faker.word.words(2);
    expected = faker.word.words(3);
    actual = faker.word.words(2);
  }

  return {
    fullName,
    email,
    feedbackType,
    appVersion,
    stepsToReproduce: steps,
    expectedBehavior: expected,
    actualBehavior: actual,
    screenshotProvided,
    label
  };
}

async function main() {
  const labels = [
    ...Array(800).fill('complete'),
    ...Array(600).fill('vague'),
    ...Array(400).fill('incomplete'),
    ...Array(200).fill('invalid')
  ] as BugReport['label'][];

  const shuffled = shuffle(labels);
  const reports = shuffled.map((label) => generateEntry(label));

  const outputPath = 'data/synthetic_bug_reports.json';
  await writeFile(outputPath, JSON.stringify(reports, null, 2));
  console.log(`Wrote ${reports.length} reports to ${outputPath}`);
  console.log('Sample:', reports.slice(0, 5));
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});

File: scripts/generate_synthetic_data.py
import json
import random
from pathlib import Path
from faker import Faker

faker = Faker()

FEEDBACK_TYPES = ["Bug", "Feature", "UI Issue"]
DEVICES = ["Android 13", "iOS 17", "Windows 11", "macOS 14", "Ubuntu 22.04"]
ACTIONS = [
    "tap the Save button",
    "open the settings page",
    "log out",
    "upload a file",
    "enter text in the search bar",
    "start the app",
    "switch accounts",
    "change the theme",
    "select a menu item",
    "resize the window",
]
ISSUES = [
    "crashes with error code 500",
    "freezes after login",
    "shows a blank screen",
    "does not respond",
    "crashed imediately",
    "bttn not responsive",
    "screen go blank",
    "resets the form",
    "closes unexpectedly",
    "shows garbled text",
]


def valid_version() -> str:
    major = random.randint(0, 5)
    minor = random.randint(0, 9)
    patch = random.randint(0, 9)
    suffix = random.choice(["", "-beta", "-rc1"])
    return f"v{major}.{minor}.{patch}{suffix}"


def invalid_version() -> str:
    return random.choice([
        "v1",
        "1.0",
        "version" + str(random.randint(6, 99)),
        f"v{faker.word()}",
        "",
    ])


def sentence_with_issue() -> str:
    device = random.choice(DEVICES)
    action = random.choice(ACTIONS)
    issue = random.choice(ISSUES)
    return f"On {device}, when I {action}, the app {issue}."


def short_issue() -> str:
    return random.choice([
        "app crashed",
        "it froze",
        "something broke",
        "not sure",
        "",
    ])


def generate_entry(label: str) -> dict:
    if label == "invalid" and random.random() < 0.3:
        full_name = f"{faker.word()} {faker.word()}"
    else:
        full_name = faker.name()

    if label == "invalid":
        email = faker.word() + ".com"
    elif label == "incomplete" and random.random() < 0.3:
        email = ""
    else:
        email = faker.email()

    feedback_type = random.choice(FEEDBACK_TYPES)
    screenshot_provided = random.choice([True, False])

    app_version = valid_version()
    steps = sentence_with_issue()
    expected = f"The app should {faker.word()} without errors."
    actual = f"Instead, it {random.choice(ISSUES)}."

    if label == "vague":
        steps = short_issue()
        expected = random.choice(["should work", "should not crash", ""])
        actual = random.choice(["didn't", "crashed", ""])
    elif label == "incomplete":
        app_version = valid_version() if random.random() < 0.5 else invalid_version()
        steps = short_issue() if random.random() < 0.5 else ""
        expected = "" if random.random() < 0.5 else "should work"
        actual = "" if random.random() < 0.5 else random.choice(["didn't", "crashed"])
    elif label == "invalid":
        app_version = invalid_version()
        steps = f"{faker.word()} {faker.word()}"
        expected = f"{faker.word()} {faker.word()} {faker.word()}"
        actual = f"{faker.word()} {faker.word()}"

    return {
        "fullName": full_name,
        "email": email,
        "feedbackType": feedback_type,
        "appVersion": app_version,
        "stepsToReproduce": steps,
        "expectedBehavior": expected,
        "actualBehavior": actual,
        "screenshotProvided": screenshot_provided,
        "label": label,
    }


def main() -> None:
    labels = (
        ["complete"] * 800
        + ["vague"] * 600
        + ["incomplete"] * 400
        + ["invalid"] * 200
    )
    random.shuffle(labels)
    reports = [generate_entry(label) for label in labels]

    output_path = Path("data/synthetic_bug_reports.json")
    output_path.write_text(json.dumps(reports, indent=2))
    print(f"Wrote {len(reports)} reports to {output_path}")
    print("Sample:", reports[:5])


if __name__ == "__main__":
    main()
